### PR TITLE
Create tree-react-api package

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add tree-react-api package.

This package provides some experimental utilities for working with SharedTree, react and DataObjects.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/framework/tree-react-api/src/test/reactSharedTreeView.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { SchemaFactory, TreeConfiguration } from "@fluidframework/tree";
+import type { ContainerSchema } from "@fluidframework/fluid-static";
+import { treeDataObject } from "../reactSharedTreeView.js";
```
Comment (CraigMacomber) [inline experimental/framework/tree-react-api/src/test/reactSharedTreeView.spec.ts:8]: This is erroring with:

> Output file '/home/craig/Work/FluidFramework.git/proxy/experimental/framework/tree-react-api/lib/reactSharedTreeView.d.ts' has not been built from source file '/home/craig/Work/FluidFramework.git/proxy/experimental/framework/tree-react-api/src/reactSharedTreeView.tsx'.ts(6305)

But I can't figure out why. The project reference from test to non-test looks correct and I'm not sure what else to check.

File: experimental/framework/tree-react-api/src/test/tsconfig.json [hunk 2]
```diff
@@ -1,20 +1,15 @@
...
-		"noImplicitAny": true,
-		"noUnusedLocals": false,
-		"noImplicitOverride": true,
 		"rootDir": "./",
 		"outDir": "../../lib/test",
-		"module": "Node16",
-		"moduleResolution": "Node16",
-		"types": ["node", "mocha"],
+		"types": ["mocha", "node"],
+		"noUnusedLocals": false,
```
Comment (tylerbutler) [inline experimental/framework/tree-react-api/src/test/tsconfig.json:7]: Consider adding a comment to this explaining why it's overridden.

Comment (CraigMacomber) [reply experimental/framework/tree-react-api/src/test/tsconfig.json]: removed

File: experimental/framework/tree-react-api/api-extractor-cjs.json
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/api-extractor-cjs.json:1]: I think I did it correctly.

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/api-extractor-cjs.json:1]: Looks like I messed something up since public.d.ts is ending up empty.

**Comment by tylerbutler** [comment]
Run `pnpm checks:fix` from the root, then `pnpm i` since some deps are outdated for the new package, then CI should pass.

File: feeds/public.txt [hunk 3]
```diff
@@ -67,6 +67,7 @@ fluid-framework
 @fluidframework/core-interfaces
 @fluidframework/container-definitions
 @fluid-internal/client-utils
+@fluid-experimental/tree-react-api
```
Comment (tylerbutler) [inline feeds/public.txt:70]: Feed files doing their job. 👍🏻 You can be confident that the package will wind up on npm when we release.

**Comment by CraigMacomber** [comment]
> pnpm checks:fix

Done, but the generate `public.d.ts` files still have no exports.

File: experimental/framework/tree-react-api/api-extractor-cjs.json
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/api-extractor-cjs.json:1]: https://github.com/microsoft/FluidFramework/pull/20780 should fix this.

File: experimental/framework/tree-react-api/.eslintrc.cjs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+module.exports = {
+	extends: [require.resolve("@fluidframework/eslint-config-fluid/strict"), "prettier"],
+	parserOptions: {
+		// TODO: why does tree package not have to list test config here?
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/.eslintrc.cjs:9]: I believe it has an `overrides` block that specifies the other config for test files.

File: experimental/framework/tree-react-api/.gitignore [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Compiled TypeScript
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/.gitignore:1]: Is this file needed? I think it can just inherit properties from the root config to capture both of these.

File: experimental/framework/tree-react-api/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+There is no service implementation agnostic client abstraction that can be referred to here (ex: shared by TinyliciousClient, AzureClient and OdspClient).
+This makes documenting compatibility with that implicit common API difficult.
+It also makes writing service agnostic code at that abstraction level harder.
+
+There does not appear to be a local service implementation of the above mentioned abstraction, which makes testing the code in the package harder.
+
+The commonly used boilerplate for setting up a ContainerSchema based application configures the dev-tools,
+but currently can't be included in this package due to dependency layering issues.
+
+<!-- AUTO-GENERATED-CONTENT:START (README_TRADEMARK_SECTION:includeHeading=TRUE) -->
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/README.md:34]: There are also section macros for contribution guidelines, general help, etc. that might be worth using. You could also use the mega `LIBRARY_PACKAGE_README` pragma to grab most everything.

Comment (Josmithr) [reply experimental/framework/tree-react-api/README.md:34]: https://github.com/microsoft/FluidFramework/blob/main/tools/markdown-magic/README.md

File: experimental/framework/tree-react-api/src/packageVersion.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+/*!
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/packageVersion.ts:1]: I don't _think_ this package needs the `packageVersion` stuff. I believe that's something needed for framework packages to determine compatibility at runtime. I don't think this package would need this.

File: experimental/framework/tree-react-api/src/reactSharedTreeView.tsx [hunk 8]
```diff
@@ -99,80 +124,90 @@ export interface ITreeDataObject<TSchema extends ImplicitFieldSchema> {
...
+ * @public
+ */
+export interface TreeViewProps<TSchema extends ImplicitFieldSchema> {
+	/**
+	 * Component to display the tree content.
+	 */
+	readonly viewComponent: React.FC<{ root: TreeFieldFromImplicitField<TSchema> }>;
+	/**
+	 * If not provided a default component is provided which describes the situation
+	 * and allows the user to upgrade the schema if possible.
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/reactSharedTreeView.tsx:141]: Nit: This could probably go into a `@defaultValue` block.

File: experimental/framework/tree-react-api/src/reactSharedTreeView.tsx [hunk 9]
```diff
@@ -43,32 +37,51 @@ const SharedTree = configuredSharedTree({
...
+			`TreeDataObject:${key}`,
+			SchemaAwareTreeDataObject,
+			[SharedTree.getFactory()],
+			{},
+		);
 	};
 }
 
+/**
+ * A schema aware Tree DataObject.
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/reactSharedTreeView.tsx:80]: Nit
```suggestion
 * A schema-aware Tree DataObject.
```

File: experimental/framework/tree-react-api/src/useTree.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * Without a way to get invalidation callbacks for specific fields,
+ * its impractical to implement an ergonomic and efficient more fiend grained invalidation hook.
+ * @public
+ */
+export function useTree(subtreeRoot: TreeNode): void {
+	// Use a React effect hook to invalidate this component when the inventory changes.
+	// We do this by incrementing a counter, which is passed as a dependency to the effect hook.
+	const [invalidations, setInvalidations] = React.useState(0);
+
+	// React effect hook that increments the 'invalidation' counter whenever inventory or any of its children change.
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/useTree.ts:23]: Docs here are out of date

Comment (Josmithr) [reply experimental/framework/tree-react-api/src/useTree.ts:23]: Also above

File: experimental/framework/tree-react-api/src/useTree.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import { Tree, type TreeNode } from "@fluidframework/tree";
+import * as React from "react";
+
+/**
+ * Custom hook which invalidates a React Component when there is a change in the subtree defined by `subtreeRoot`.
+ * This includes changes to the tree's content, but not changes to its parentage.
+ * See {@link @fluidframework/tree#TreeChangeEvents.treeChanged} for details.
+ * @privateRemarks
+ * Without a way to get invalidation callbacks for specific fields,
+ * its impractical to implement an ergonomic and efficient more fiend grained invalidation hook.
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/useTree.ts:15]:
```suggestion
 * it's impractical to implement an ergonomic and efficient more fiend grained invalidation hook.
```

File: experimental/framework/tree-react-api/src/useTree.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ */
+export function useTree(subtreeRoot: TreeNode): void {
+	// Use a React effect hook to invalidate this component when the inventory changes.
+	// We do this by incrementing a counter, which is passed as a dependency to the effect hook.
+	const [invalidations, setInvalidations] = React.useState(0);
+
+	// React effect hook that increments the 'invalidation' counter whenever inventory or any of its children change.
+	React.useEffect(() => {
+		// Returns the cleanup function to be invoked when the component unmounts.
+		return Tree.on(subtreeRoot, "treeChanged", () => {
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/useTree.ts:26]: I'm not following this logic. Why are we subscribing to `Tree.on` when the component unmounts? Shouldn't this subscribe when the effect is run, and unsubscribe in the cleanup function?

Comment (CraigMacomber) [reply experimental/framework/tree-react-api/src/useTree.ts:26]: we are unsubscribing from tree.on when it unmounts. I think you read it wrong.

File: experimental/framework/tree-react-api/.gitignore (hunk 5 above)
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/.gitignore]: removed

File: experimental/framework/tree-react-api/src/packageVersion.ts (hunk 7 above)
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/src/packageVersion.ts]: Correct. Removed (along with the package script to generate it)

File: experimental/framework/tree-react-api/src/useTree.ts (hunk 12 above)
Comment (Josmithr) [reply experimental/framework/tree-react-api/src/useTree.ts:26]: Oh, `Tree.on` returns a cleanup function. I see.

File: experimental/framework/tree-react-api/api-report/tree-react-api.api.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    readonly errorComponent?: React_2.FC<SchemaIncompatibleProps>;
+    readonly viewComponent: React_2.FC<{
+        root: TreeFieldFromImplicitField<TSchema>;
+    }>;
+}
+
+// @public
+export function useTree(subtreeRoot: TreeNode): void;
+
+// (No @packageDocumentation comment for this package)
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/api-report/tree-react-api.api.md:56]: We should probably add a `@packageDocumentation` comment in `index.ts` so there is a package summary in the API docs.

File: packages/framework/fluid-static/src/types.ts [hunk 14]
```diff
@@ -46,11 +46,16 @@ export type LoadableObjectClass<T extends IFluidLoadable = IFluidLoadable> =
  *
  * @typeParam T - The class of the `DataObject`.
  * @privateRemarks
- * Having both `factory` and `LoadableObjectCtor` is redundant, and having `factory` not actually work as a factory is also strange.
- * This may need some refinement.
+ * Having both `factory` and `LoadableObjectCtor` is redundant.
+ * TODO: It appears the factory is whats used, so the constructor should be removed.
```
Comment (Josmithr) [inline packages/framework/fluid-static/src/types.ts:50]:
```suggestion
 * TODO: It appears the factory is what's used, so the constructor should be removed.
```

**Review by Josmithr — APPROVED** [review]

File: experimental/framework/tree-react-api/api-report/tree-react-api.api.md (hunk 13 above)
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/api-report/tree-react-api.api.md:56]: would be real slick if markdown magic could inject the package comment into the readme to avoid duplication.
